src/lzw.ml

```diff
@@ -11,58 +20,75 @@ let rec get_bits bytedata offset count =
 
 let update_buffer buffer offset codestr =
-  List.fold_left
-    (fun acc c ->
-      Cstruct.set_char buffer acc c;
-      acc + 1)
-    offset codestr
+  let len = String.length codestr in
+  Unsafe.blit_string codestr 0 buffer offset len;
+  offset + len
+
+let max_dict_size = 4096
+
+let init_table () =
+  Array.init max_dict_size (function
+    | i when i < 256 -> String.make 1 (Char.chr i)
+    | _ -> String.empty)
+
+let reset_table dict =
+  Array.iteri
+    (fun i _ ->
+      Array.set dict i
+        (if i < 256 then String.make 1 (Char.chr i) else String.empty))
+    dict
 
 let decode input output_buffer =
-  let dict : lzwstring array = Array.make 4096 [] in
+  let dict : lzwstring array = init_table () in
 
-  let rec inner in_offset out_offset code_size next_code_index prev_code =
+  let rec inner in_offset out_offset code_size next_code_index
+      (prev_code : lzwstring) =
     let code = get_bits input in_offset code_size in
     match code with
     | 257 -> () (* end code *)
     | 256 -> (
         (* clear code *)
+        reset_table dict;
         let next_code = get_bits input (in_offset + code_size) 9 in
         match next_code with
         | 257 -> ()
         | _ ->
-            let out_offset =
-              update_buffer output_buffer out_offset
-                [ Char.chr (next_code land 0xFF) ]
-            in
-            inner
-              (in_offset + code_size + 9)
-              out_offset 9 258
-              [ Char.chr (next_code land 0xFF) ])
+            let next = String.make 1 (Char.chr (next_code land 0xFF)) in
+            let out_offset = update_buffer output_buffer out_offset next in
+            inner (in_offset + code_size + 9) out_offset 9 258 next)
     | _ ->
         let entry =
-          if code < 256 then [ Char.chr (code land 0xFF) ]
-          else if code < next_code_index then dict.(code)
-          else build_table_entry prev_code prev_code
+          if code < next_code_index then (
+            let entry = dict.(code) in
+            if next_code_index < max_dict_size then
+              dict.(next_code_index) <-
+                build_table_entry ~previous:prev_code entry;
+            entry)
+          else
+            let e = build_table_entry ~previous:prev_code prev_code in
+            if next_code_index < max_dict_size then dict.(next_code_index) <- e;
+            e
         in
 
-        if next_code_index < 4096 then
-          dict.(next_code_index) <- build_table_entry prev_code entry;
         let prev_entry, new_code_size, new_code_index =
           match next_code_index with
           | 4095 -> (prev_code, code_size, next_code_index)
           | _ ->
-              let i = next_code_index + 1 in
-              ( entry,
-                (if i >= (1 lsl code_size) - 1 then code_size + 1 else code_size),
-                i )
+              let new_code_size =
+                match next_code_index with
+                | 510 | 1022 | 2046 -> code_size + 1
```

Comment:
If I'm being picky then these magic numbers could use a comment (but I'm also guilty of that)

Reply:
Indeed!